File: 076_minimum_window_substring.java
class Solution {
    public String minWindow(String s, String t) {
        if (t.length() > s.length()) return "";

        Map<Character, Integer> map = new HashMap<>();
        for (char c: t.toCharArray()) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        int counter = map.size();

        int begin = 0, end = 0;
        int head = 0;
        int len = Integer.MAX_VALUE;

        while (end < s.length()) {
            char c = s.charAt(end);
            if (map.containsKey(c)) {
                map.put(c, map.get(c) - 1);
                if (map.get(c) == 0) counter--;
            }
            end++;

            while (counter == 0) {
                char tc = s.charAt(begin);
                if (map.containsKey(tc)) {
                    map.put(tc, map.get(tc) + 1);
                    if (map.get(tc) > 0) {
                        counter++;
                    }
                }
                if (end - begin < len) {
                    len = end - begin;
                    head = begin;
                }
                begin++;
            }
        }
        if (len == Integer.MAX_VALUE) return "";
        return s.substring(head, head + len);
    }
}

File: python/076_minimum_window_substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        map = dict(collections.Counter(t))
        counter = len(map)

        head = begin = end = 0
        min_len = float('inf')
        while end < len(s):
            temp = s[end]
            if temp in map:
                map[temp] = map[temp] - 1
                if map[temp] == 0:
                    counter -= 1

            end += 1

            while counter == 0:
                temp_b = s[begin]
                if temp_b in map:
                    map[temp_b] = map[temp_b] + 1
                    if map[temp_b] > 0:
                        counter += 1
                if end - begin < min_len:
                    min_len = end - begin
                    head = begin
                begin += 1

        if min_len == float('inf'): return ""
        return s[head:head + min_len]
